sdk/python/kfp/compiler/compiler.py

```diff
@@ -59,16 +61,29 @@ def my_pipeline(a: int = 1, b: str = "default value"):
 
   def __init__(
       self,
-      mode: dsl.PipelineExecutionMode = dsl.PipelineExecutionMode.V1_LEGACY,
+      mode: Optional[dsl.PipelineExecutionMode] = None,
       launcher_image: Optional[str] = None):
     """Creates a KFP compiler for compiling pipeline functions for execution.
 
     Args:
-      mode: The pipeline execution mode to use.
+      mode: The pipeline execution mode to use, defaults to V1_LEGACY.
+        KF_PIPELINES_COMPILER_MODE env var can override the default. For example,
+        KF_PIPELINES_COMPILER_MODE=V2_COMPATIBLE.
       launcher_image: Configurable image for KFP launcher to use. Only applies
         when `mode == dsl.PipelineExecutionMode.V2_COMPATIBLE`. Should only be
         needed for tests or custom deployments right now.
     """
+    if mode is None:
+      mode_str = os.environ.get(KF_PIPELINES_COMPILER_MODE_ENV, 'V1_LEGACY')
```

Comment:
Feels to me that allowing environmental variable is like opening a backdoor that breaks the interface. Env variable can be set or unset from anywhere and at anytime. The same code could behave differently on different machines or on the same machine but at different times. It could be a nightmare for debuggability. 
To me, the cons of using env variable outweighs the pros.

That being said, I'm okay if we want to use it here, just for this compatibility setting, given it's a transient setting which will go away when KFP v2 is ready, but not extending this pattern to other arguments.

Reply:
I agree that supporting env vars or not is a trade off between explicitness and convenience.

I have a different viewpoint on what matters in this scenario, remember that `dsl-compile` is a CLI tool. A CLI tool is used very often on different targets. It would be a lot more inconvenient if you use an argument all the time, but you have to type it each time. For the same reason, `kubectl`, `gcloud`, `go` etc CLI tools allow configuring some persistent configuration through env vars and/or config file, so that when you use them in a terminal. You'd only need to type the information related to your current command.

Allowing implicit configuration does make it harder to debug. So we need to spend more effort on this, e.g. `go env` prints out all your go environment variables, so that you can understand configuration for your current context. For our case, I've been using `env | grep KF` pretty happily to confirm my current setup. Maybe we can make that a formal command to make it easier to understand.

FYI, this is not the first introduction of env vars
https://github.com/kubeflow/pipelines/blob/8b9756247c68b681fc9a7c3ea1ce0dab5e22c847/sdk/python/kfp/_client.py#L84-L90
KFP client has supported configuration from env vars [since 0.1.25](https://github.com/kubeflow/pipelines/commit/6e5171452b923ba2da381d182a7d4e313d9f5321).

Another benefit of env vars, is that it can help decouple deployment specific configuration from pipeline/run specific configuration. e.g. if we document how users can configure env vars to let KFP client know how to talk to the API endpoint, then our samples do not need to always leave a placeholder for endpoint that people need to change before running. This can create an experience of always using the same command to run samples, see feature request https://github.com/kubeflow/pipelines/issues/5381

Reply:
Thanks for the detailed explanation. That makes sense. 
Given what you described, I would still prefer we have this env read logic in the CLI implementation (sdk/python/kfp/compiler/main.py), rather than inside the compiler (sdk/python/kfp/compiler/compiler.py). So that we can limit such pattern to cli/client use case only.

That said, it's still not a big deal for this particular argument.

Reply:
Good point, only adding env read logic in CLI implementation does feel better and enough for UX!

Reply:
Finished the implementation change